File: src/core/pairing.js
import net from 'node:net';
import crypto from 'node:crypto';
import { EventEmitter } from 'node:events';
import { PROTOCOL_VERSION, loadConfig, platformLabel } from './config.js';
import { FrameChannel } from './protocol.js';
import { publicKeyString, importPublicKey, sharedSecret } from './identity.js';
import { rememberPeer } from './trust.js';

const COMMIT_LABEL = 'flyshare-sas-commit-v2';
const SAS_LABEL = 'flyshare-sas-v2';
const DECISION_TIMEOUT_MS = 3 * 60 * 1000;
const STEP_TIMEOUT_MS = 30 * 1000;

/**
 * Six-digit short authentication string.
 *
 * Both machines derive it from the same X25519 exchange, so a person comparing
 * the two screens is checking that nothing sits in the middle: an attacker
 * relaying the pairing ends up with a different shared secret on each leg, and
 * therefore a different number on each screen.
 */
export function computeSasCode({ initiatorPublic, responderPublic, initiatorNonce, responderNonce, shared }) {
  const salt = Buffer.concat([b64(initiatorNonce), b64(responderNonce)]);
  const info = Buffer.from(`${SAS_LABEL}|${initiatorPublic}|${responderPublic}`, 'utf8');
  const material = Buffer.from(crypto.hkdfSync('sha256', shared, salt, info, 4));
  return String(material.readUInt32BE(0) % 1000000).padStart(6, '0');
}

/**
 * The responder commits to its public key and nonce before seeing the
 * initiator's.
 *
 * Without this step the scheme is broken, not merely weaker: a machine in the
 * middle picks its own key pairs, so it could try key after key until the two
 * codes it shows happen to match — about a million cheap attempts. Committing
 * first means neither side can still be shopping for a key once it knows what
 * the other one chose, which puts an attacker back at one chance in a million.
 */
export function pairingCommitment(publicKey, nonce) {
  return crypto.createHash('sha256')
    .update(COMMIT_LABEL)
    .update(publicKey)
    .update(b64(nonce))
    .digest('base64url');
}

function b64(value) {
  return Buffer.from(value, 'base64url');
}

function newNonce() {
  return crypto.randomBytes(16).toString('base64url');
}

function selfDescriptor() {
  const config = loadConfig();
  return { id: config.deviceId, name: config.deviceName, os: platformLabel() };
}

/**
 * Dial a peer and run the pairing exchange.
 * `onCode` fires as soon as the code is known, so the UI can show it while the
 * far side is still waiting for a person to press the button.
 */
export function initiatePairing(peer, onCode) {
  const self = selfDescriptor();
  const myPublic = publicKeyString();
  const myNonce = newNonce();

  return new Promise((resolve, reject) => {
    const socket = net.connect({ host: peer.address, port: peer.port });
    let settled = false;

    const fail = (err) => {
      if (settled) return;
      settled = true;
      socket.destroy();
      reject(err);
    };
    const done = (value) => {
      if (settled) return;
      settled = true;
      socket.end();
      resolve(value);
    };

    socket.setTimeout(STEP_TIMEOUT_MS, () => fail(new Error('the other device stopped responding')));
    socket.once('error', (err) => fail(new Error(`cannot reach ${peer.name} — ${err.message}`)));
    socket.once('connect', async () => {
      try {
        const channel = new FrameChannel(socket);
        channel.write({ t: 'pair', ver: PROTOCOL_VERSION, device: self });

        const commitFrame = await expect(channel, 'pair-commit');
        channel.write({ t: 'pair-reveal', pub: myPublic, nonce: myNonce });

        const open = await expect(channel, 'pair-open');
        if (pairingCommitment(open.pub, open.nonce) !== commitFrame.commit) {
          throw new Error('the other device changed its key mid-pairing — pairing aborted');
        }

        const peerKey = importPublicKey(open.pub);
        const code = computeSasCode({
          initiatorPublic: myPublic,
          responderPublic: open.pub,
          initiatorNonce: myNonce,
          responderNonce: open.nonce,
          shared: sharedSecret(peerKey),
        });
        onCode?.(code);

        // The person now compares the two screens and answers on the far side.
        socket.setTimeout(DECISION_TIMEOUT_MS);
        const result = await expect(channel, 'pair-result', DECISION_TIMEOUT_MS);
        if (!result.accept) throw new Error(result.reason ?? 'the other device declined');

        const device = open.device ?? { id: peer.id, name: peer.name, os: peer.os };
        rememberPeer({ id: device.id, name: device.name, os: device.os, publicKey: open.pub });
        done({ ...device, code });
      } catch (err) {
        fail(err);
      }
    });
  });
}

/** Read the next frame, insisting it is the one the protocol calls for. */
async function expect(channel, type, timeoutMs = STEP_TIMEOUT_MS) {
  const frame = await channel.read(timeoutMs);
  if (frame.t === 'pair-error') throw new Error(frame.reason ?? 'pairing failed');
  if (frame.t !== type) throw new Error(`unexpected reply "${frame.t}" during pairing`);
  return frame;
}

/**
 * The receiving half of a pairing exchange, held open while a person decides.
 * Mirrors IncomingTransfer: the UI sees it, shows the code, and answers.
 */
export class IncomingPairing extends EventEmitter {
  #decided = false;
  #timer = null;

  constructor(socket, hello, seed) {
    super();
    this.socket = socket;
    this.channel = new FrameChannel(socket, seed);
    this.id = crypto.randomUUID();
    this.peer = hello.device ?? { id: 'unknown', name: 'unknown device', os: 'unknown' };
    this.code = null;
    this.status = 'exchanging';
    this.error = null;
    this.createdAt = Date.now();

    this.publicKey = publicKeyString();
    this.nonce = newNonce();

    socket.once('error', () => this.#fail('connection lost'));
    socket.once('close', () => this.#fail('the other device disconnected'));
  }

  /** Run the commit / reveal / open exchange, then park until answered. */
  async run() {
    try {
      this.socket.setTimeout(STEP_TIMEOUT_MS, () => this.#fail('the other device stopped responding'));
      this.#send({ t: 'pair-commit', commit: pairingCommitment(this.publicKey, this.nonce) });

      const reveal = await expect(this.channel, 'pair-reveal');
      this.peerPublic = reveal.pub;
      this.peerNonce = reveal.nonce;

      this.#send({
        t: 'pair-open',
        pub: this.publicKey,
        nonce: this.nonce,
        device: selfDescriptor(),
      });

      this.code = computeSasCode({
        initiatorPublic: reveal.pub,
        responderPublic: this.publicKey,
        initiatorNonce: reveal.nonce,
        responderNonce: this.nonce,
        shared: sharedSecret(importPublicKey(reveal.pub)),
      });

      this.status = 'awaiting-confirmation';
      this.socket.setTimeout(0);
      this.#timer = setTimeout(() => this.decide(false, 'no answer'), DECISION_TIMEOUT_MS);
      this.emit('change');
    } catch (err) {
      this.#fail(err.message);
    }
  }

  decide(accept, reason) {
    if (this.#decided) return;
    this.#decided = true;
    clearTimeout(this.#timer);

    if (accept) {
      rememberPeer({
        id: this.peer.id,
        name: this.peer.name,
        os: this.peer.os,
        publicKey: this.peerPublic,
      });
      this.status = 'paired';
    } else {
      this.status = 'rejected';
      this.error = reason ?? 'declined on this device';
    }
    this.#send({ t: 'pair-result', accept, reason });
    this.emit('change');
    setTimeout(() => this.socket.end(), 100);
  }

  #send(obj) {
    this.channel.write(obj);
  }

  #fail(message) {
    if (this.#decided || this.status === 'paired') return;
    this.#decided = true;
    clearTimeout(this.#timer);
    this.status = 'failed';
    this.error = message;
    this.socket.destroy();
    this.emit('change');
  }

  snapshot() {
    return {
      id: this.id,
      kind: 'pairing',
      direction: 'in',
      peer: this.peer,
      code: this.code,
      status: this.status,
      error: this.error,
      createdAt: this.createdAt,
    };
  }
}

File: android/core/src/main/kotlin/com/lunov/flyshare/core/Pairing.kt
package com.lunov.flyshare.core

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.Socket

/**
 * The pairing exchange, both roles — docs/PROTOCOL.md §7.
 *
 * Written against streams rather than sockets so the whole handshake can be
 * driven over a pipe in a unit test, with no network involved.
 */
object Pairing {

    const val STEP_TIMEOUT_MS = 30_000
    const val DECISION_TIMEOUT_MS = 180_000

    /** What the far side told us about itself. Display data, not identity. */
    data class RemoteDevice(val id: String, val name: String, val os: String)

    data class Outcome(val peer: PairedPeer, val code: String)

    private fun deviceObject(self: SelfDescription): JsonObject = buildJsonObject {
        put("id", JsonPrimitive(self.id))
        put("name", JsonPrimitive(self.name))
        put("os", JsonPrimitive(self.os))
    }

    private fun readDevice(frame: JsonObject): RemoteDevice? {
        val device = frame.obj("device") ?: return null
        val id = device.string("id") ?: return null
        return RemoteDevice(id, device.string("name") ?: id, device.string("os") ?: "unknown")
    }

    /**
     * Dial a peer and run the exchange. `onCode` fires the moment the number is
     * known, so it can go on screen while the far side is still waiting for a
     * person to press a button.
     */
    fun initiate(
        address: String,
        port: Int,
        self: SelfDescription,
        identity: Identity,
        trust: TrustStore,
        onCode: (String) -> Unit,
    ): Outcome {
        Socket().use { socket ->
            socket.tcpNoDelay = true
            socket.connect(InetSocketAddress(address, port), STEP_TIMEOUT_MS)
            socket.soTimeout = STEP_TIMEOUT_MS
            return initiateOver(socket.getInputStream(), socket.getOutputStream(), self, identity, trust, onCode) {
                socket.soTimeout = DECISION_TIMEOUT_MS
            }
        }
    }

    internal fun initiateOver(
        input: InputStream,
        output: OutputStream,
        self: SelfDescription,
        identity: Identity,
        trust: TrustStore,
        onCode: (String) -> Unit,
        beforeDecision: () -> Unit = {},
    ): Outcome {
        val myPublic = identity.publicKeyString
        val myNonce = Crypto.randomBytes(16).toBase64Url()

        Frames.write(output, frame("t" to "pair", "ver" to PROTOCOL_VERSION, "device" to deviceObject(self)))

        val commitFrame = expect(input, "pair-commit")
        val commit = commitFrame.string("commit") ?: throw PairingException("commitment was missing")

        Frames.write(output, frame("t" to "pair-reveal", "pub" to myPublic, "nonce" to myNonce))

        val open = expect(input, "pair-open")
        val peerPublic = open.string("pub") ?: throw PairingException("peer sent no public key")
        val peerNonce = open.string("nonce") ?: throw PairingException("peer sent no nonce")

        // The commitment is the whole point: without checking it here, a device
        // in the middle could pick its key after seeing ours and grind the two
        // displayed codes into agreement.
        if (Sas.commitment(peerPublic, peerNonce) != commit) {
            throw PairingException("the other device changed its key mid-pairing — pairing aborted")
        }

        val shared = Crypto.x25519(identity.privateKey, peerPublic.fromBase64Url())
        val code = Sas.code(myPublic, peerPublic, myNonce, peerNonce, shared)
        onCode(code)

        beforeDecision()
        val result = expect(input, "pair-result")
        if (result.string("accept") != "true") {
            throw PairingException(result.string("reason") ?: "the other device declined")
        }

        val device = readDevice(open) ?: throw PairingException("peer did not identify itself")
        val peer = PairedPeer(device.id, device.name, device.os, peerPublic, System.currentTimeMillis())
        trust.remember(peer)
        return Outcome(peer, code)
    }

    /**
     * The answering half, held open while a person decides.
     *
     * `confirm` is called with the code once both sides can compute it, and
     * blocks until someone answers — that wait is the security control, not an
     * inconvenience to be engineered away.
     */
    fun respond(
        input: InputStream,
        output: OutputStream,
        hello: JsonObject,
        self: SelfDescription,
        identity: Identity,
        trust: TrustStore,
        confirm: (peer: RemoteDevice, code: String) -> Boolean,
    ): Outcome {
        val device = readDevice(hello) ?: throw PairingException("peer did not identify itself")

        val myPublic = identity.publicKeyString
        val myNonce = Crypto.randomBytes(16).toBase64Url()

        // Committed before the initiator's key is visible — §7.1.
        Frames.write(output, frame("t" to "pair-commit", "commit" to Sas.commitment(myPublic, myNonce)))

        val reveal = expect(input, "pair-reveal")
        val peerPublic = reveal.string("pub") ?: throw PairingException("peer sent no public key")
        val peerNonce = reveal.string("nonce") ?: throw PairingException("peer sent no nonce")

        Frames.write(
            output,
            frame(
                "t" to "pair-open",
                "pub" to myPublic,
                "nonce" to myNonce,
                "device" to deviceObject(self),
            ),
        )

        val shared = Crypto.x25519(identity.privateKey, peerPublic.fromBase64Url())
        val code = Sas.code(peerPublic, myPublic, peerNonce, myNonce, shared)

        val accepted = confirm(device, code)
        Frames.write(
            output,
            frame(
                "t" to "pair-result",
                "accept" to accepted,
                "reason" to if (accepted) null else "declined on the other device",
            ),
        )
        if (!accepted) throw PairingException("declined on this device")

        val peer = PairedPeer(device.id, device.name, device.os, peerPublic, System.currentTimeMillis())
        trust.remember(peer)
        return Outcome(peer, code)
    }

    private fun expect(input: InputStream, type: String): JsonObject {
        val frame = Frames.read(input)
        if (frame.type() == "pair-error") {
            throw PairingException(frame.string("reason") ?: "pairing failed")
        }
        if (frame.type() != type) {
            throw PairingException("unexpected reply \"${frame.type()}\" during pairing")
        }
        return frame
    }
}

class PairingException(message: String) : Exception(message)
